DOCSP-43384: Troubleshooting openssl legacy unsafe renegotiation (#115)
mongodb/docs-pymongo: source/includes/troubleshooting/tls.rst

- Downgrade Python to v3.9 or earlier
- Upgrade {+mdb-server+} to v4.2 or later
- Install {+driver-short+} with the :ref:`OCSP <pymongo-disable-ocsp>` option, which relies on PyOpenSSL

Unsafe Legacy Renegotiation Disabled
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When using OpenSSL v3 or later, you might see an error similar to the following
message:

.. code-block:: python

   MongoServerSelectionError: 886E0000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:c:\ws\deps\openssl\openssl\ssl\statem\extensions.c:922:

> These types of errors occur due to outdated or buggy SSL proxies that mistakenly

Reviewer: I: [due to](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#d:~:text=number%20of%20years.-,due%20to,-Use%20because%20of)

```suggestion
These types of errors occur because of outdated or buggy SSL proxies that mistakenly
```

enforce legacy `TLS renegotiation <https://www.ibm.com/docs/en/i/7.3?topic=settings-renegotiation>`__. 

> To resolve this issue, use the ``UnsafeLegacyServerConnect`` option with the 
> ``OPENSSL_CONF`` environment variable. Create a configuration
> file with the following content:

Reviewer: I: this intro sentence jumps right to the steps, making it seem like they're actually two different steps. maybe try something like this:

```suggestion
To resolve this issue, create a configuration file that includes the
``UnsafeLegacyServerConnect`` option, as shown in the following code example:
```

you could also add [highlighting ](https://www.mongodb.com/docs/meta/reference/code-blocks/#syntax)to the line with the relevant option

.. code-block:: shell

   openssl_conf = openssl_init

   [openssl_init]
   ssl_conf = ssl_sect

   [ssl_sect]
   system_default = system_default_sect

   [system_default_sect]
   Options = UnsafeLegacyServerConnect

Then run Python using that OpenSSL config file:

.. code-block:: shell

>    OPENSSL_CONF=/path/to/the/config/file/above.cnf python ...

Reviewer: Q: this syntax is unusual to me. are you setting an environment variable and also starting python? setting the variable within the context of this run of python?

Author: also: see [here](https://askubuntu.com/questions/1421436/whats-the-difference-between-and-space-when-setting-environment-variable-be) - confirmed, only setting the variable within this run of python

Reviewer: thanks!

The ``UnsafeLegacyServerConnect`` option in ``OPENSSL_CONF`` requires OpenSSL v3.0.4 

> or greater.

Reviewer: S: Move to earlier, when you first mention this option

> .. warning::

Reviewer: S: other teams use "warning" to indicate possible data loss, so i try to follow suit and use "important" instead. also, what's the reason for not using this option? unsafe?

```suggestion
.. important::
```

- Downgrade Python to v3.9 or earlier
- Upgrade {+mdb-server+} to v4.2 or later
- Install {+driver-short+} with the :ref:`OCSP <pymongo-disable-ocsp>` option, which relies on PyOpenSSL

Unsafe Legacy Renegotiation Disabled
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When using OpenSSL v3 or later, you might see an error similar to the following
message:

.. code-block:: python

   MongoServerSelectionError: 886E0000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:c:\ws\deps\openssl\openssl\ssl\statem\extensions.c:922:

These types of errors occur because of outdated or buggy SSL proxies that mistakenly
enforce legacy `TLS renegotiation <https://www.ibm.com/docs/en/i/7.3?topic=settings-renegotiation>`__. 

> To resolve this issue, create a configuration file that includes the 
> ``UnsafeLegacyServerConnect`` option. This option requires OpenSSL v3.0.4 or 
> greater. The following example shows how to set the ``UnsafeLegacyServerConnect`` 
> option:

Reviewer: S: mentioning the dependency after the first instruction feels out-of-order. what if they ensure have the dependency met first, then make the file?

also: [Style Guide on earlier/later](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#e:~:text=Use%20these%20terms%20(instead%20of%20lower%20and%20higher%20or%20below%20and%20above)%20when%20referring%20to%20product%20releases%20and%20version%20numbers.)

```suggestion
To resolve this issue, perform the following steps:

1. Run the following command to ensure that you have OpenSSL v3.0.4 or later installed:

   .. code-block:: bash
   
      openssl version
      
#. Create a configuration file that includes the 
``UnsafeLegacyServerConnect`` option.  The following example shows how to set the ``UnsafeLegacyServerConnect`` 
option:
```

Running python could then be step 3. You can also use the [`procedure` directive](https://www.mongodb.com/docs/meta/style-guide/style/tasks/#procedures) if you think it looks better!

   [system_default_sect]
   Options = UnsafeLegacyServerConnect

> Then run Python while setting the ``OPENSSL_CONF`` environment variable to use

Reviewer: ```suggestion
Then, run Python while setting the ``OPENSSL_CONF`` environment variable to use
```

the OpenSSL configuration file you just created:

.. code-block:: shell

   OPENSSL_CONF=/path/to/the/config/file/above.cnf python ...

.. important::

   Because setting the ``UnsafeLegacyServerConnect`` option has 
   `security implications <https://docs.openssl.org/3.0/man3/SSL_CTX_set_options/#patched-openssl-client-and-unpatched-server>`__, 

>    this workaround should only be used as a last 
>    resort to address ``unsafe legacy renegotiation disabled`` errors.

Reviewer: I: avoid "should" + put "only" in the right place + avoid passive voice

```suggestion
   use this workaround only as a last 
   resort to address ``unsafe legacy renegotiation disabled`` errors.
```
